[PK-244] Adding type checking for setRange/setFilter

packpaul opened the pull request:
## Overview

Please describe your changes here.

---

### Checklist

- [x] Change is covered by automated tests.
- [x] JavaDoc is provided for all new/changed public APIs.
- [ ] User Guide is updated.
- [x] CI builds pass.

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
Refactoring: Column -> Column<?>;
Some Ru->En translations for comments provided

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
ColumnNamed -> ColumnRef, and adapted for ASC/DESC in Column.orderBy()

packpaul commented:
I think that ``ru.curs.celesta.dbutils.BasicCursor.BasicCursor(CallContext, Set<String>)`` should have its counterpart - ``ru.curs.celesta.dbutils.BasicCursor.BasicCursor(CallContext, Set<ColumnRef<?>>)``.

packpaul commented on celesta-sql/src/main/java/ru/curs/celesta/score/BinaryColumn.java:
```diff
@@ -5,7 +5,7 @@
  * Binary column (IMAGE or BLOB type).
  */
-public final class BinaryColumn extends Column {
+public final class BinaryColumn extends Column<String> {
```
I'm slightly in doubt which generic type should be used here ``<String>`` or ``<BLOB>``, therefore I decided not to include column refs into generated cursors for binary type columns.

inponomarev replied:
Funny thing is that it really doesn't matter. We need type parameter for Column for type-checked versions of setRange/setFilter, but we cannot set a range or a filter on BLOB field.
So -- really doesn't matter

packpaul commented on celesta-sql/src/main/java/ru/curs/celesta/score/Column.java:
```diff
@@ -116,7 +118,7 @@ public final void delete() throws ParseException {
      * @return
      */
-    public abstract Object getDefaultValue();
+    public abstract V getDefaultValue();
```
Generic parameter works well here except for ``BinaryColumn`` type (see my comment above).

inponomarev replied:
...and that's because Binary Column cannot be used in `setRange`, so it's OK)

packpaul commented on celesta-sql/src/main/java/ru/curs/celesta/dbutils/adaptors/ddl/DdlGenerator.java:
```diff
@@ -205,13 +205,16 @@ String createTable(TableElement te) {
-    final String columnDef(Column c) {
+    final String columnDef(Column<?> c) {
+        @SuppressWarnings("unchecked")
+        final Class<? extends Column<?>> cClass = (Class<Column<?>>) c.getClass();
```
I don't know how to make it without the excplicit type conversion.

inponomarev replied:
Doesn't look a big problem to me. To avoid `@SuppressWarnings`,  we might just use `Class<? extends Column>` all the way down instead of `Class<? extends Column<?>>`...

packpaul commented on celesta-core/src/main/java/ru/curs/celesta/dbutils/BasicCursor.java:
```diff
@@ -329,6 +330,15 @@ static BasicCursor create(DataGrainElement element, CallContext callContext, Set
         return (Class<? extends BasicCursor>) Class.forName(cursorClassName, true, Thread.currentThread().getContextClassLoader());
     }
 
+    protected static <T> ColumnRef<T> createColumnReference(final String columnName) {
```
Is ``ColumnRef`` naming OK?

inponomarev replied:
as  I wrote you in Telegram we might rethink the thing )

packpaul commented on celesta-maven-plugin/src/main/java/ru/curs/celesta/plugin/maven/CursorGenerator.java:
```diff
@@ -314,6 +317,21 @@ private static String calcClassName(GrainElement ge, String sourceFileNamePrefix
+                .filter(e -> e.getValue().getJavaClass() != BLOB.class)
+                .map(e -> FieldSpec.builder(
+                        ParameterizedTypeName.get(ColumnRef.class, e.getValue().getJavaClass()),
+                        e.getKey() + "_COLUMN",
```
Instead of **_COLUMN** one could have **_REF**, **_FIELD** or something else. I'm biased, but I think that **_COLUMN** is not the best choice either.

inponomarev replied:
In generated classes it looks rather ugly and not Java style. Why not just add `...Column` instead, like `...Cursor`?

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
Soothing checkstyle

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
ColumnRef<V> merged with ColumnMeta<V>;
*Cursor.COLUMN.* provided for column referencing

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
ViewColumnMeta.setName() for Views added

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
IT fixed

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
Checkstyles fixes

packpaul commented on celesta-maven-plugin/src/main/java/ru/curs/celesta/plugin/maven/CursorGenerator.java:
```diff
@@ -314,8 +320,40 @@ private static String calcClassName(GrainElement ge, String sourceFileNamePrefix
+    }
+
+    private static FieldSpec buildColumnsField(TypeName columnsClassType) {
+        return FieldSpec.builder(columnsClassType, "COLUMNS", Modifier.PUBLIC, Modifier.FINAL)
```
I made the field name **COLUMNS** capitalized in order to escape probable naming conflicts when there's a column named **columns**.

inponomarev reviewed (changes requested):
1. For now, new versions of `setRange`, `orderBy` etc. are calling the deprecated versions of the methods. This is incorrect and should be made the other way round: the deprecated methods should call new ones in a way that allows all the deprecated methods be simply deleted some day.

Here is an idea: all of the 'legacy' methods are using `void validateColumnName`, which simply throws an exception in case there is no column with such name. We can make this method return the needed ColumnMeta and then this returned ColumnMeta can be used in 'new methods', like:

```java
@Deprecated         
    @SuppressWarnings("unchecked")
    public final void setRange (String name, Object value) {
        setRange((ColumnMeta<Object>) validateColumnName(name), value);
}
```
 (and when we get rid of the legacy versions of the methods, `validateColumName` will also be deleted)
2. The implementation of the 'fellow' `...Columns` classes do not follow the points that we discussed on Telegram. Let me repeat them once again:

* This object should be static, and thus we should be able to instatiate it separately from the Cursor and without any `CallContext` available.
* It should take `Celesta` as its constructor parameter, and in its constructor, it should extract 
and store in the final field the correct meta-object (during the code generation phase we know the name of the table, so we can code-generate the correct call chain to `Celesta.getScore..getGrain..getTable`)
* In order to save memory (we will have one fellow object per each cursor!) we decided to have methods, not final fields, to extract the `ColumnMeta`s.

3. The implementation of point #2 will allow us to do the following:

* introduce `public BasicCursor(CallContext context, Set<ColumnMeta<?>> fields)` constructor
* deprecate the `public BasicCursor(CallContext context, Set<String> fields)` constructor

All this will lead us to statically checked functionality of `setRange`, `setFilter`, `orderBy`, and cursor instatiation in one PR.

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
*Cursor.*Columns -> static *Cursor.Columns with column fields moved to
methods

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
Cursor.Cursor(сontext, columns), ViewCursor.Cursor(сontext, columns),
ParameterizedViewCursor.Cursor(сontext, parameters, columns) introduced

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
BasicCursor deprecated methods inverted to use up-to-date methods

inponomarev commented:
Woo-hoo, the implementation is perfect AFAICS!! Thank you )

Just one final thing: there are no **tests** that check the direct call of new versions of `setRange`/`setFilter`. I believe we should add some tests in `celesta-test`/`ru.curs.celesta.script` (a simple thing: instantiate one of the already-existing test table cursors, add a couple of records and play with the outcome of `setRange`/`setFilter` calls). By this we will also get the examples of the new API usage.

There are also tests that use deprecated versions of `setRange` (e. g. search for `_testInFilterWithRangeInMainCursor`), and they cannot be easily rewritten, because we need the dynamic approach there. On the second thought, I think that for now we can leave these tests as they are. Maybe we will rewrite them later, maybe we will cancel the deprecation of 'dynamic' `setRange`. We should try to start using these new APIs anyway.

We will also need to update the **user guide**. I can do it if you like.

packpaul commented:
@inponomarev 
It'd be nice if you could overtake the documentation part, since I'm running out of time for the other tasks. Indeed I haven't added any special tests for the new methods but I thought changing old versions of them in other tests would be enough. I'll add special tests then.

packpaul committed:
[PK-244] Adding type checking for setRange/setFilter
FieldsLookup.add(ColumnMeta, ColumnMeta) added;
Explicit tests for Cursor.setRange and Cursor.setFilter

packpaul commented:
@inponomarev May I kindly ask you to do a review again?

inponomarev reviewed (approved):
:thumbsup: 

LGTM now. Thank you very much!

inponomarev commented:
Now I'm going to update the User Guide and then merge the PR
